src/main/scala/io/citrine/lolo/trees/splits/AnnealingSplitter.scala

```diff
@@ -0,0 +1,196 @@
+        case x: Char => AnnealingSplitter.getBestCategoricalSplit(data, totalSum, totalWeight, index, minInstances, beta)
+        case x: Any => throw new IllegalArgumentException("Trying to split unknown feature type")
+      }
+    }
+
+    val totalProbability = possibleSplits.map(_._3).sum
+    val draw = Random.nextDouble() * totalProbability
+    var cumSum: Double = 0.0
+    possibleSplits.foreach { case (split, variance, probability) =>
+      cumSum = cumSum + probability
+      if (draw < cumSum) {
+        val deltaImpurity = -variance - totalSum * totalSum / totalWeight
+        return (split, deltaImpurity)
+      }
+    }
+    throw new RuntimeException("Draw was beyond all the probabilities")
+  }
+}
+
+object AnnealingSplitter {
+  /**
+    * Find the best split on a continuous variable
+    *
+    * @param data        to split
+    * @param totalSum    Pre-computed data.map(d => data._2 * data._3).sum
+    * @param totalWeight Pre-computed data.map(d => d._3).sum
+    * @param index       of the feature to split on
+    * @return the best split of this feature
+    */
+  def getBestRealSplit(
+                        data: Seq[(Vector[AnyVal], Double, Double)],
+                        totalSum: Double,
+                        totalWeight: Double,
+                        index: Int,
+                        minCount: Int,
+                        beta: Double
+                      ): (RealSplit, Double, Double) = {
+    /* Pull out the feature that's considered here and sort by it */
+    val thinData = data.map(dat => (dat._1(index).asInstanceOf[Double], dat._2, dat._3)).sortBy(_._1)
+
+    /* Base cases for iteration */
+    var leftSum = 0.0
+    var leftWeight = 0.0
+
+    /* Move the data from the right to the left partition one value at a time */
+    val possibleSplits: Seq[(Double, Double, Double)] = (0 until data.size - minCount).map { j =>
+      leftSum = leftSum + thinData(j)._2 * thinData(j)._3
+      leftWeight = leftWeight + thinData(j)._3
+
+      /* This is just relative, so we can subtract off the sum of the squares, data.map(Math.pow(_._2, 2)) */
+      val totalVariance = -leftSum * leftSum / leftWeight - Math.pow(totalSum - leftSum, 2) / (totalWeight - leftWeight)
+
+      /* Keep track of the best split, avoiding splits in the middle of constant sets of feature values
+         It is really important for performance to keep these checks together so
+         1) there is only one branch and
+         2) it is usually false
+       */
+      val score: Double = if (j + 1 >= minCount && thinData(j + 1)._1 > thinData(j)._1 + 1.0e-9) {
+        Math.exp(-totalVariance * beta)
+      } else {
+        0.0
+      }
+      val pivot = (thinData(j + 1)._1 + thinData(j)._1) / 2.0
+      (score, pivot, totalVariance)
+    }
+
+    val totalScore = possibleSplits.map(_._1).sum
+    val draw = Random.nextDouble() * totalScore
+    var cumSum: Double = 0.0
+    possibleSplits.foreach{case (score, pivot, variance) =>
```

Comment:
I like this logic.  Adding a comment explaining how it works could help future generations.

Reply:
Added.  I'd like to DRY it out with a helper too, but I don't think that's low hanging fruit right now.